io-sim/src/Control/Monad/IOSimPOR/Internal.hs

```diff
@@ -0,0 +1,1573 @@
+  let wakeup = ordNub [ tid | (_vid, tids) <- tidss, tid <- reverse tids ]
+      wokeby = Map.fromListWith Set.union
+                                [ (tid, Set.singleton vid)
+                                | (vid, tids) <- tidss
+                                , tid <- tids ]
+  return (wakeup, wokeby)
+
+ordNub :: Ord a => [a] -> [a]
+ordNub = go Set.empty
+  where
+    go !_ [] = []
+    go !s (x:xs)
+      | x `Set.member` s = go s xs
+      | otherwise        = x : go (Set.insert x s) xs
+
+-- Effects
+
+data Effect = Effect {
+    effectReads  :: !(Set TVarId),
+    effectWrites :: !(Set TVarId),
+    effectForks  :: !(Set ThreadId),
+    effectLiftST :: !Bool,
+    effectThrows :: ![ThreadId],
+    effectWakeup :: ![ThreadId]
+  }
+  deriving (Eq, Show)
+
+instance Semigroup Effect where
+  Effect r w s b ts wu <> Effect r' w' s' b' ts' wu' =
+    Effect (r<>r') (w<>w') (s<>s') (b||b') (ts++ts') (wu++wu')
+
+instance Monoid Effect where
+  mempty = Effect Set.empty Set.empty Set.empty False [] []
+
+-- readEffect :: SomeTVar s -> Effect
+-- readEffect r = mempty{effectReads = Set.singleton $ someTvarId r }
+
+readEffects :: [SomeTVar s] -> Effect
+readEffects rs = mempty{effectReads = Set.fromList (map someTvarId rs)}
+
+-- writeEffect :: SomeTVar s -> Effect
+-- writeEffect r = mempty{effectWrites = Set.singleton $ someTvarId r }
+
+writeEffects :: [SomeTVar s] -> Effect
+writeEffects rs = mempty{effectWrites = Set.fromList (map someTvarId rs)}
+
+forkEffect :: ThreadId -> Effect
+forkEffect tid = mempty{effectForks = Set.singleton $ tid}
+
+liftSTEffect :: Effect
+liftSTEffect = mempty{ effectLiftST = True }
+
+throwToEffect :: ThreadId -> Effect
+throwToEffect tid = mempty{ effectThrows = [tid] }
+
+wakeupEffects :: [ThreadId] -> Effect
+wakeupEffects tids = mempty{effectWakeup = tids}
+
+someTvarId :: SomeTVar s -> TVarId
+someTvarId (SomeTVar r) = tvarId r
+
+onlyReadEffect :: Effect -> Bool
+onlyReadEffect e@Effect { effectReads } = e == mempty { effectReads }
+
+racingEffects :: Effect -> Effect -> Bool
+racingEffects e e' =
+      (effectLiftST e  && racesWithLiftST e')
+   || (effectLiftST e' && racesWithLiftST e )
+   || (not $ null $ effectThrows e `List.intersect` effectThrows e')
+   || (not $ effectReads  e `Set.disjoint` effectWrites e'
+          && effectWrites e `Set.disjoint` effectReads  e'
+          && effectWrites e `Set.disjoint` effectWrites e')
+  where racesWithLiftST eff =
+             effectLiftST eff
+          || not (Set.null (effectReads eff) && Set.null (effectWrites eff))
+
+-- Steps
+
+data Step = Step {
+    stepThreadId :: !ThreadId,
+    stepStep     :: !Int,
+    stepEffect   :: !Effect,
+    stepVClock   :: !VectorClock
+  }
+  deriving Show
+
+-- steps race if they can be reordered with a possibly different outcome
+racingSteps :: Step -> Step -> Bool
+racingSteps s s' =
+     stepThreadId s /= stepThreadId s'
+  && not (stepThreadId s' `elem` effectWakeup (stepEffect s))
```

Comment:
I don't think so. racingSteps is applied to an earlier step and a later step, and it must return True it is is interesting and possible to reschedule the earlier step after the later one. It's only interesting if the effects interfere, of course, but it's only _possible_ if the steps are in different threads, and the later step's thread was not "woken up" by the earlier step. If that was the case, then trying to schedule the later step before the earlier one results in trying to schedule a thread that is not running, which is impossible. 

Now, what about the symmetric case, what if the later step, in its first execution, woke up the thread that performed the earlier step (which may have gone to sleep in the meantime)? Well, that's neither here nor there... it doesn't mean we can't reverse the steps (although if we do, then the second step, when it is repeated, will not wake the first thread up, because it is already running).

If we did add the symmetric condition to this function, then it would prevent us from exploring some feasible schedules, and possibly result in bugs being missed.

Reply:
Right, what was I thinking about? The reordering of events is clearly not a symmetric operation. It would probably help me when reading, to just document
```hs
racingSteps :: Step -- ^ earlier step
            -> Step -- ^ later step
            -> Bool
```